updated documentation to new format (#214) (#28)
eclipse-dataspaceconnector/MinimumViableDataspace: system-tests/README.md

# System tests

The system tests copies a file from a provider to a consumer blob storage account.

## Publish/Build Tasks

> ! Important Note !
> 
> MVD depends on Eclipse DataSpaceConnector(EDC), Identity Hub and Registration Service. These dependencies 
> are __not__ published to any central artifact repository yet, so in local development we have to use locally 
> published dependencies.
>

> >In order to use the correct version of each repo required by the `MVD`, you need to look in [action.yml](./.github/actions/../../../.github/actions/gradle-setup/action.yml) for the hashes of the versions of the `EDC`, `Identity Hub` and the `Registration Service` that are being used by the `MVD`.

Reviewer: why the paths with `..` (also in other places)

>
> For Example, the `Registration Service`, `Identity Hub` and `EDC` repository hash can be found in the _Checkout_ steps  (in the `ref` property) of [action.yml](./.github/actions/../../../.github/actions/gradle-setup/action.yml):

```yml
    - name: Checkout EDC
      uses: actions/checkout@v2
      with:
        repository: eclipse-dataspaceconnector/DataSpaceConnector
        path: DataSpaceConnector

>         ref: 3ff940b720f44826df28e893fb31344eb6faacef

Reviewer: should we remove these values so that people don't believe they should get them from here

Author: The instructions indicate it is an example and show where to get them from; the `ref` value could be removed, or updated as needed until the time that it is not longer needed (when dependencies are published)

Reviewer: Yes I got that, I just expect that we will forget to update them / users will wrongly take them from here, hence I think it's better to remove these values from here

    - name: Checkout Registration Service
      uses: actions/checkout@v2
      with:
        repository: eclipse-dataspaceconnector/RegistrationService
        path: RegistrationService
        ref: 374c14bcca23ddb1dcd7476a27264510e54de7fa

    - name: Checkout Identity Hub
      uses: actions/checkout@v2
      with:
        repository: eclipse-dataspaceconnector/IdentityHub
        path: IdentityHub
        ref: bc13cf0cb8589b792eef733c7cf7b3422476add5

```

> After you have cloned the `EDC`, `Identity Hub` and `Registration Service` repos locally you should run the command to
> `reset` to the specific hash.  
>
> For Example:

```bash
# EDC (in the EDC root folder)

> git reset --hard 3ff940b720f44826df28e893fb31344eb6faacef

Reviewer: ```suggestion
git checkout 3ff940b720f44826df28e893fb31344eb6faacef
```
also in other places

Author: will make this change but it forces a detached head state.  coming with next PR

# Identity Hub (in the Identity Hub root folder)
git reset --hard bc13cf0cb8589b792eef733c7cf7b3422476add5

# Registration Service (in the Registration Service root folder)
git reset --hard 374c14bcca23ddb1dcd7476a27264510e54de7fa
```

> Now you can follow the rest of the process below.  

> > Once the publications are available in _Maven Central_ this process will not be necessary

Reviewer: ```suggestion
```

This is potentially confusing for the user, I would remove it (or rephrase it to something clearer, like "It is planned...")

Author: not sure what is potentially confusing in this statement.

Reviewer: The user might not be aware of plans to publish artefacts in Maven Central, by whom and when?

> 
<br />

### EDC

<br />

Execute the following command  from `EDC` root folder.

```bash
./gradlew publishToMavenLocal -Pskip.signing
```

_Note for Windows PowerShell, the following command should be used:_

```powershell

> ./gradlew publishToMavenLocal -P"skip.signing"

Reviewer: `./gradlew publishToMavenLocal -P skip.signing`
would work on both platforms

Reviewer 2: On windows PowerShell it complains due to dot(.)

```

<br />

### Identity Hub

<br />

Execute the following command from `Identity Hub` root folder:

```bash
./gradlew publishToMavenLocal -Pskip.signing
```

_Note for Windows PowerShell, the following command should be used:_

```powershell
./gradlew publishToMavenLocal -P"skip.signing"
```

<br />

### Registration Service

<br />

Execute the following command from `Registration Service` root folder:

```bash
./gradlew publishToMavenLocal
```

<br />

### MVD

<br />

Now that the publishing to the local repositories has been completed, `MVD` can be built by running the following command from the root of the `MVD` project folder:

```bash

> ./gradlew build -x test

Reviewer: why should tests be skipped?

Reviewer 2: Otherwise integration tests will fail as docker is still not started as per instructions.

If we want to run it with test then all integration tests needs to be disabled first by default. (Can be a seperate PR)

Author: not making any change to this PR for this; as @cpeeyush indicated seperate PR could address

```

## Local Test Execution

- `MVD` system tests can be executed locally against a local `MVD` instance. 
- `MVD` runs three `EDC Connectors` and one `Registration Service`.

_Note: Ensure that you are able to build `MVD` locally as described in the previous [section](#mvd)._

> First, we need to build `EDC Connector`, `Identity Hub` and the `Registration Service`.

Reviewer: ```suggestion
First, we need to build the `EDC Connector` (which also includes the `Identity Hub`) and `Registration Service` runtimes.
```

From the `MVD` root folder, execute the following command:

```bash
./gradlew -DuseFsVault="true" :launcher:shadowJar
```

From the `Registration Service` root folder, execute the following command:

```bash
./gradlew :launcher:shadowJar
```

From the `MVD` root folder execute the following commands to set the `Registration Launcher` path environment variable and start `MVD` using the `docker-compose.yml` file.

```bash

> export REGISTRATION_SERVICE_LAUNCHER_PATH=/home/user/RegistrationService/launcher

Reviewer: should we explain that the path should be adapted

Author: can resolve in next PR for the document as there are a number of other changes coming in to reflect MVD changes

docker-compose -f system-tests/docker-compose.yml up --build
```

_Note for Windows PowerShell, the following commands should be used from the `MVD` project root.  (The path will depend on the location of the your `RegistrationService` project root):_

```powershell
$Env:REGISTRATION_SERVICE_LAUNCHER_PATH="c:/RegistrationService/launcher"
docker-compose -f system-tests/docker-compose.yml up --build
```

Once completed, following services will start within their docker containers:
- 3 `EDC Connectors`
  - _consumer-us_
  - _consumer-eu_
  - _provider_ (which will also be seeded with initial required data using a [postman collection](../deployment/data/MVD.postman_collection.json))
- A `Registration Service`

> - A `HTTP Nginx Server` (to serve DIDs) 

Reviewer: ```suggestion
- A `HTTP Nginx Server` (to serve DID Documents)
```

- An `Azurite` blob storage service 
  

_Note, the `Newman` docker container will automatically stop after seeding initial data from postman scripts._

`EDC Connectors` need to be registered using `Registration Service` CLI client jar. After publishing `Registration Service` locally the client jar should be available under the `Registration Service` root project folder in _client-cli/build/libs_.

```bash
# Replace path according to your local set up
export REGISTRATION_SERVICE_CLI_JAR_PATH=c:/RegistrationService/client-cli/build/libs/registration-service-cli.jar

# Register Participants

> ./system-tests/resources/register-participants.sh

Reviewer: note that @cpeeyush is changing this in https://github.com/agera-edc/MinimumViableDataspace/pull/178 (work in progress), will need to be updated

```

_Note for Windows PowerShell, the following commands should be run the the `MVD` root project folder._

```powershell
# Replace path according to your local set up

$Env:REGISTRATION_SERVICE_CLI_JAR_PATH = "c:\RegistrationService\client-cli\build\libs\registration-service-cli.jar"

# Register Provider

> java -jar $Env:REGISTRATION_SERVICE_CLI_JAR_PATH -s="http://localhost:8184/api" participants add --request="{ \`"name\`": \`"provider\`", \`"supportedProtocols\`": [ \`"ids-multipart\`" ], \`"url\`": \`"http://provider:8282\`" }"

Reviewer: note that @Izzzu is changing this, will need to be updated

Author: then it can be addressed in a follow up

# Register Consumer-EU
java -jar $Env:REGISTRATION_SERVICE_CLI_JAR_PATH -s="http://localhost:8184/api" participants add --request="{ \`"name\`": \`"consumer-eu\`", \`"supportedProtocols\`": [ \`"ids-multipart\`" ], \`"url\`": \`"http://consumer-eu:8282\`" }"

# Register Consumer-US
java -jar $Env:REGISTRATION_SERVICE_CLI_JAR_PATH -s="http://localhost:8184/api" participants add --request="{ \`"name\`": \`"consumer-us\`", \`"supportedProtocols\`": [ \`"ids-multipart\`" ], \`"url\`": \`"http://consumer-us:8282\`" }"
```

Set the environment variable `TEST_ENVIRONMENT` to `local` to enable local blob transfer test and then run `MVD` system test using the following command:

```bash
export TEST_ENVIRONMENT=local
./gradlew :system-tests:test
```

_Note for Windows PowerShell, the following commands should be used:_

```powershell
$Env:TEST_ENVIRONMENT = "local"
./gradlew :system-tests:test
```

> > [Storage Explorer](https://azure.microsoft.com/features/storage-explorer/) can be used to connect to the `Azurite` storage container on `127.0.0.1:10000` port and under `consumereuassets`, account transferred blob can be viewed.

Reviewer: ```suggestion
> [Storage Explorer](https://azure.microsoft.com/features/storage-explorer/) can be used to connect to the `Azurite` storage container on `127.0.0.1:10000` port and under the `consumereuassets` account, the transferred blob can be viewed.
```

> > [Storage Explorer](https://azure.microsoft.com/features/storage-explorer/) can be used to connect to the `Azurite` storage container on `127.0.0.1:10000` port and under `consumereuassets`, account transferred blob can be viewed.

Reviewer: Is `consumereuassets` really the name of the destination account (`transferred blob` seems to imply)? It sounds more like the source account

Author: @cpeeyush your thoughts on this?  I did not alter this section from the original

Reviewer 2: yes, it is the name of the storage account for consumer-eu. Here is the list of others https://github.com/eclipse-dataspaceconnector/MinimumViableDataspace/blob/e8721ba0e4e7177ebba2dad865cb8536a745f753/system-tests/docker-compose.yml#L133.

The naming is one similar lines(not exact) as we have in the cloud deployment.

### Local Test Resources

The following test resources are provided in order to run `MVD` locally. `system-tests/docker-compose.yml` uses it to start `MVD`.

<br>

---

<br>

> Each `EDC Connector` has its own set of Private and Public keys with java keystore e.g. `system-tests/resources/provider`. These were generated using the following commands:

Reviewer: ```suggestion
Each `EDC Connector` has its own set of Private and Public keys in PEM and Java KeyStore formats, e.g. `system-tests/resources/provider`. These were generated using the following commands:
```

```bash
# generate a private key
openssl ecparam -name prime256v1 -genkey -noout -out private-key.pem
# generate corresponding public key
openssl ec -in private-key.pem -pubout -out public-key.pem
# create a self-signed certificate
openssl req -new -x509 -key private-key.pem -out cert.pem -days 360
```

> Generated keys are imported to keystores e.g. `system-tests/resources/provider/provider-keystore.jks`. Each keystore has password `test123`.

Reviewer: path includes a `vault` component, e.g. `system-tests/resources/vault/provider/` (in multiple places)

        path: RegistrationService
        ref: 374c14bcca23ddb1dcd7476a27264510e54de7fa

    - name: Checkout Identity Hub
      uses: actions/checkout@v2
      with:
        repository: eclipse-dataspaceconnector/IdentityHub
        path: IdentityHub
        ref: bc13cf0cb8589b792eef733c7cf7b3422476add5

```

> After you have cloned the `EDC`, `Identity Hub` and `Registration Service` repos locally you should run the command to
> `checkout` to the specific hash.  
>
> For Example:

```bash
# EDC (in the EDC root folder)
git checkout 3ff940b720f44826df28e893fb31344eb6faacef

# Identity Hub (in the Identity Hub root folder)
git checkout bc13cf0cb8589b792eef733c7cf7b3422476add5

# Registration Service (in the Registration Service root folder)
git checkout 374c14bcca23ddb1dcd7476a27264510e54de7fa
```

> Now you can follow the rest of the process below.  
> Once the publications are available in _Maven Central_ this process will not be necessary
> 
<br />

### EDC

<br />

Execute the following command  from `EDC` root folder.

```bash
./gradlew publishToMavenLocal -Pskip.signing
```

_Note for Windows PowerShell, the following command should be used:_

```powershell

> ./gradlew publishToMavenLocal -P"skip.signing"

Author: addressed

```

<br />

### Identity Hub

<br />

Execute the following command from `Identity Hub` root folder:

```bash
./gradlew publishToMavenLocal -Pskip.signing
```

_Note for Windows PowerShell, the following command should be used:_

```powershell
./gradlew publishToMavenLocal -P"skip.signing"
```

<br />

### Registration Service

<br />

Execute the following command from `Registration Service` root folder:

```bash
./gradlew publishToMavenLocal
```

<br />

### MVD

<br />

Now that the publishing to the local repositories has been completed, `MVD` can be built by running the following command from the root of the `MVD` project folder:

```bash
./gradlew build -x test
```

## Local Test Execution

- `MVD` system tests can be executed locally against a local `MVD` instance. 
- `MVD` runs three `EDC Connectors` and one `Registration Service`.

_Note: Ensure that you are able to build `MVD` locally as described in the previous [section](#mvd)._

First, we need to build the `EDC Connector` (which also includes the `Identity Hub`) and `Registration Service` runtimes.

From the `MVD` root folder, execute the following command:

```bash

> ./gradlew -DuseFsVault="true" :launcher:shadowJar

Author: addressed

./gradlew :system-tests:test
```

> [Storage Explorer](https://azure.microsoft.com/features/storage-explorer/) can be used to connect to the `Azurite` storage container on `127.0.0.1:10000` port and under the `consumereuassets` account, the transferred blob can be viewed.

### Local Test Resources

The following test resources are provided in order to run `MVD` locally. `system-tests/docker-compose.yml` uses it to start `MVD`.

<br>

---

<br>

Each `EDC Connector` has its own set of Private and Public keys in PEM and Java KeyStore formats, e.g. `system-tests/resources/vault/provider`. These were generated using the following commands:

```bash
# generate a private key
openssl ecparam -name prime256v1 -genkey -noout -out private-key.pem
# generate corresponding public key
openssl ec -in private-key.pem -pubout -out public-key.pem
# create a self-signed certificate
openssl req -new -x509 -key private-key.pem -out cert.pem -days 360
```

Generated keys are imported to keystores e.g. `system-tests/resources/vault/provider/provider-keystore.jks`. Each keystore has password `test123`.

> [KeyStore Explorer](https://keystore-explorer.org/) can be used to manage keystores from UI.

> `MVD` local instance usage `EDC File System Vault` and its keys are managed using a java properties file e.g.`system-tests/resources/vault/provider/provider-vault.properties`.

Reviewer 3: ```suggestion
`MVD` local instances use a file-system based vault and its keys are managed using a java properties file e.g.`system-tests/resources/vault/provider/provider-vault.properties`.
```
